### ISSUE TITLE
Mapping bs-seq samples to multiple specific reference fasta

### ISSUE THREAD

**Issue by rbpisupati** [nf-core/methylseq#181]
Hi,

I was thinking of adapting this pipeline to take care of multiple reference genomes. Something like each sample would be aligned to different reference fasta file. Any votes for such a feature? 

The input `if given` would be a csv file with sample_id, path_to_reference_fasta, path_to_sample_fastq

Cheers, 
Rahul

**Comment by phue** [comment]
I think this would be a nice addition, particularly for population (epi)genetics.
Something similar to a design file as was done over at nf-core/chipseq by @drpatelh 

We need to think about how this could work together with iGenomes though

**Comment by drpatelh** [comment]
I have done something similar in the nanoseq pipeline.  Check out the format of the input samplesheet [here](https://nf-co.re/nanoseq/1.1.0/usage#introduction). The important thing is to have a validation script for the samplesheet that checks the entries provided by users. I wrote [this](https://github.com/nf-core/nanoseq/blob/ad5b2bb7a37a6edb9f61603636bd26e43259f58b/main.nf#L304) function to resolve the iGenomes per sample if provided. Note, you will also have to configure the channels to only build genome indices once if multiple samples share the same reference. It will be tricky but not impossible 🙂

**Comment by phue** [comment]
There is now a [dsl2 branch](https://github.com/nf-core/methylseq/tree/dsl2) with this functionality.

Please give it a try :+1:

**Comment by ewels** [comment]
NB: This was removed from the dsl2 branch so that we could work with centralised modules in nf-core/modules.

However, I then wrote `nf-core modules patch` to allow local modifications of remote DSL2 modules for exactly this use case ☝🏻 

Now it just needs implementing again..

**Comment by mobilegenome** [comment]
It would be great to get this issue going.  For my population epigenomics project, I will have an individual genome for each sample, so I would priotize providing paths to FASTA files in the '`genome` column. Also, I'm happy to chip in work but would appreciate if you could help with some starting points. If I understand correctly there has been some previous work, that's still in the history. 

Second, do I understand correctly that would like to keep the nf-core modules untouched so they stay for general use?

@njspix do you have a full repository of the pipeline with your additions available and mind sharing it?

**Comment by Emiller88** [comment]
So is this:
1.
```csv
sample,fastq_1,fastq_2,genome
SAMPLE_PAIRED_END,/path/to/fastq/files/AEG588A1_S1_L002_R1_001.fastq.gz,/path/to/fastq/files/AEG588A1_S1_L002_R2_001.fastq.gz,hg19
SAMPLE_SINGLE_END,/path/to/fastq/files/AEG588A4_S4_L003_R1_001.fastq.gz,,hg18
```
or 

2. Just supporting something like `nextflow run --fasta hg18,hg19` and run it on all the samples.

The first I think is going to be fundamentally incompatible with nf-core modules for now (See https://github.com/nextflow-io/nextflow/issues/2085). Nextflow doesn't join across metas and there's just no way to link between tuples like you'd expect.

The second option might be pretty easy.

I think in either case, it might just be easier to run the pipeline twice, but I get the desire to want to do so.

**Comment by njspix** [comment]
Hey @Emiller88 could you clarify why 1) is incompatible with nf-core modules?? I'm not following you 100%

**Comment by njspix** [comment]
ok - thank you so much for clarifying!! 
If i'm understanding correctly, this is the problem that my suggestion [above](https://github.com/nf-core/methylseq/issues/181#issuecomment-1308882396) was designed to address. 

It's a bit magical (and I don't mean that in a good way), but it allows you to use the pattern
```
    input:
    tuple val(meta), path(reads)
    path index
```
_**as is**_ by synchronizing the two input channels. in essence, you `join` the reads channel and the index channel, then `multiMap` them into two sub-channels. These can then be passed to the aligner separately (e.g. `align(reads, indexes)` but will **stay in sync** with each other.

let me know if i'm missing the bus here! 

Thanks much for your work on this.

**Comment by Emiller88** [comment]
> If i'm understanding correctly, this is the problem that my suggestion https://github.com/nf-core/methylseq/issues/181#issuecomment-1308882396 was designed to address.

Ah, I missed that, that's pretty cool! 

I guess we can just add it to a meta map, but in the methylseq pipeline, I don't actually see any other use of the reference besides just alignment? I just searched for "fasta", "index" "genome" and it only looks like it's the first few steps. So that would be pretty easy to maintain!

**Comment by njspix** [comment]
Yeah it's a pretty nifty feature! I'm not sure about `bismark` and other aligners, but I know that `biscuit` utilizes the reference (not the index, though) for several steps, including bisulfite conversion filtering, pileup, and merging CpGs. 

It's a bit inelegant but my first attempt would probably be something like this for each process requiring a reference:
```
process1.out.map{meta, bam -> [meta.genome, meta, bam]}
    .combine(refs, by: 0)
    .multiMap{ genome, meta, bam, ref ->
        bams: [meta, bam]
        refs: ref
    }.set{ for_process2 }
    
process2(for_process2.bams, for_process2.refs)
```

If that's too clunky I wonder if we could abstract out the repeated code into a function...

**Comment by njspix** [comment]
Good to know, thanks very much!
